### PR TITLE
808 add pgrouting dataset type to stack data uploader

### PR DESCRIPTION
**Pull request by cmclscourtney**
Adds a `OSMRouting` data-subset type to the `stack-data-uploader`. An simple example would be the following with `routing` directory containing a `.osm` and `.xml` config file.

```json
{
    "database": "postgres",
    "workspace": "the_world_avatar",
    "datasetDirectory": "routing",
    "dataSubsets": [
        {
            "type": "OSMRouting",
            "skip": false,
            "schema": "public",
            "name": "my_routing",
            "osm2PGRoutingOptions": {
                "flags": [
                    "--attributes"
                ],
                "options": {
                    "--chunk": "40000"
                }
            }
        }
    ]
}
```

### CODE REVIEW

**Comment by cmclscourtney** [comment]
@ShinZert I still need to write documentation and fix some issues with the osm2pgrouting options but spinning up the stack and uploading basic data should work

**Comment by gpeb2** [comment]
Note to self:

- [ ] Check status of underlying PostGIS "bundled" image (currently a beta version).
- [ ] Add the 3DCityDB extension to the PostGIS.

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/postgis/PostGISClient.java [hunk 1]
```diff
@@ -77,4 +108,122 @@ public RemoteRDBStoreClient getRemoteStoreClient(String database) {
...
+    private void ensurePostGISRoutingSupportEnabled(String database, String postGISContainerId) {
+        ByteArrayOutputStream errorStream = new ByteArrayOutputStream();
+        String execId = createComplexCommand(postGISContainerId,
+                "psql", "-U", postgreSQLEndpoint.getUsername(), "-d", database, "-w")
+                .withHereDocument(
+                        "CREATE EXTENSION IF NOT EXISTS pgrouting CASCADE; CREATE EXTENSION IF NOT EXISTS hstore;")
+                .withErrorStream(errorStream)
+                .exec();
+        handleErrors(errorStream, execId, logger);
+    }
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/postgis/PostGISClient.java:228]: Might it make sense to move this to a new class, for example ```PGRoutingClient```?

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/utils/FileUtils.java [hunk 2]
```diff
@@ -127,4 +127,13 @@ public static String fixSlashs(String path, boolean leading, boolean trailing) {
         return path;
     }
 
+    public static boolean hasFileExtension(Path file, String extension) {
+        int i = file.toString().lastIndexOf(".");
+        if (i > 0) {
+            return file.toString().substring(i + 1).equals(extension);
+        } else {
+            return false;
+        }
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/utils/FileUtils.java:136]: Suggested simplification:
```suggestion
        file.toString().endsWith("." + extension);
```

**Review by gpeb2 — CHANGES_REQUESTED** [review]
I've made some suggestions, I don't think any of them are crucial though.

**Review by gpeb2 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/OSMRouting.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            virtualTable.setSql(handleFileValues(virtualTable.getSql()));
+        }
+
+        GeoServerClient.getInstance()
+                .createPostGISLayer(workspaceName, database, layerName, geoServerVectorSettings);
+    }
+
+    private String getTablePrefix() {
+        return getName() + "_";
+    }
```
[the remark this answers is not in the document]
Comment (cmclscourtney) [reply Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/OSMRouting.java:54]: Resolved by [519c95b](https://github.com/cambridge-cares/TheWorldAvatar/pull/812/commits/519c95b8fe66a3ee71a8d28413139e54b3caa239)

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/geoserver/UpdatedGSVirtualTableEncoder.java [hunk 4]
```diff
@@ -18,4 +21,8 @@ public void setKeyColumn(String keycolumn) {
     public void setGeometry(VTGeometryEncoder geomEnc) {
         addVirtualTableGeometry(geomEnc);
     }
+
+    public void setParameters(List<VTParameterEncoder> paramEncList) {
+        paramEncList.stream().forEach(paramEnc -> addVirtualTableParameter(paramEnc));
```
[the remark this answers is not in the document]
Comment (cmclscourtney) [reply Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/geoserver/UpdatedGSVirtualTableEncoder.java]: Resolved by bf40099

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/utils/FileUtils.java (hunk 2 above)
Comment (cmclscourtney) [reply Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/utils/FileUtils.java]: Resolved by 8060bd9

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/postgis/PostGISClient.java (hunk 1 above)
Comment (cmclscourtney) [reply Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/postgis/PostGISClient.java]: Resolved by 8060bd9

**Review by gpeb2 — APPROVED** [review]
Looks fine to me now.

**Pull request merged**